src/main/java/com/google/devtools/build/lib/bazel/bzlmod/ModuleFileGlobals.java

```diff
@@ -651,11 +664,83 @@ public void useRepo(
     hadNonModuleCall = true;
     Location location = thread.getCallerLocation();
     for (String arg : Sequence.cast(args, String.class, "args")) {
-      extensionProxy.addImport(arg, arg, location);
+      extensionProxy.addImport(arg, arg, "by a use_repo() call", location);
     }
     for (Map.Entry<String, String> entry :
         Dict.cast(kwargs, String.class, String.class, "kwargs").entrySet()) {
-      extensionProxy.addImport(entry.getKey(), entry.getValue(), location);
+      extensionProxy.addImport(entry.getKey(), entry.getValue(), "by a use_repo() call", location);
+    }
+  }
+
+  @StarlarkMethod(
+      name = "use_repo_rule",
+      doc =
+          "Returns a proxy value that can be directly invoked in the MODULE.bazel file as a"
+              + " repository rule, one or more times. Repos created in such a way are only visible"
+              + " to the current module, under the name declared using the <code>name</code>"
+              + " attribute on the proxy. The implicit Boolean <code>dev_dependency</code>"
+              + " attribute can also be used on the proxy to denote that a certain repo is only to"
+              + " be created when the current module is the root module.",
+      parameters = {
+        @Param(
+            name = "repo_rule_bzl_file",
+            doc = "A label to the Starlark file defining the repo rule."),
+        @Param(
+            name = "repo_rule_name",
+            doc =
+                "The name of the repo rule to use. A symbol with this name must be exported by the"
+                    + " Starlark file."),
+      },
+      useStarlarkThread = true)
+  public RepoRuleProxy useRepoRule(String bzlFile, String ruleName, StarlarkThread thread) {
+    hadNonModuleCall = true;
+    ModuleExtensionUsageBuilder newUsageBuilder =
+        new ModuleExtensionUsageBuilder(
+            "//:MODULE.bazel",
+            ModuleExtensionId.INNATE_EXTENSION_NAME,
+            /* isolate= */ false,
+            Location.fromFile(thread.getCallerLocation().file()));
```

Comment:
This line confused me at first, could you add a comment explaining why we don't pass in the full location?

Reply:
added a comment